File: frontend/lib/models/route_result.dart
import 'package:latlong2/latlong.dart';

class RouteResult {
  final String? shelterId;
  final List<List<LatLng>> segments;
  final int totalDistanceM;
  final double estimatedTimeMin;
  final double totalRiskScore;

  const RouteResult({
    this.shelterId,
    required this.segments,
    required this.totalDistanceM,
    required this.estimatedTimeMin,
    required this.totalRiskScore,
  });

  bool get hasRoute => segments.isNotEmpty;

  factory RouteResult.fromJson(Map<String, dynamic> json) {
    final geoJson = json['route_geojson'] as Map<String, dynamic>?;
    final features = (geoJson?['features'] as List?) ?? [];

    final segments = <List<LatLng>>[];
    for (final feature in features) {
      final geometry =
          (feature as Map<String, dynamic>)['geometry'] as Map<String, dynamic>?;
      if (geometry?['type'] != 'LineString') continue;
      final seg = <LatLng>[];
      for (final coord in (geometry?['coordinates'] as List?) ?? []) {
        if (coord is List && coord.length >= 2) {
          seg.add(LatLng(
            (coord[1] as num).toDouble(),
            (coord[0] as num).toDouble(),
          ));
        }
      }
      if (seg.length >= 2) segments.add(seg);
    }

    return RouteResult(
      shelterId: (json['shelter'] as Map<String, dynamic>?)?['id']?.toString(),
      segments: segments,
      totalDistanceM: (json['total_distance_m'] as num?)?.toInt() ?? 0,
      estimatedTimeMin: (json['estimated_time_min'] as num?)?.toDouble() ?? 0,
      totalRiskScore: (json['total_risk_score'] as num?)?.toDouble() ?? 0,
    );
  }
}

File: frontend-rn/src/types/route-result.ts
import type { LngLat } from '@/lib/geo';

export interface RouteResult {
  shelterId?: string;
  segments: LngLat[][];          // road segments (solid line)
  connectorSegments: LngLat[][]; // GPS↔road and road↔shelter legs (dashed line)
  totalDistanceM: number;
  estimatedTimeMin: number;
  totalRiskScore: number;
  riskPercent: number;           // peak danger along route, 0–100
}

export const hasRoute = (r: RouteResult | null) => !!r && r.segments.length > 0;

export function routeResultFromJson(j: any): RouteResult {
  const features = j?.route_geojson?.features ?? [];
  const segments: LngLat[][] = [];
  const connectorSegments: LngLat[][] = [];

  for (const f of features) {
    const g = f?.geometry;
    if (g?.type !== 'LineString') continue;
    const seg: LngLat[] = (g.coordinates ?? [])
      .filter((c: any) => Array.isArray(c) && c.length >= 2)
      .map((c: any) => [Number(c[0]), Number(c[1])] as LngLat);
    if (seg.length < 2) continue;
    if (f?.properties?.connector === true) connectorSegments.push(seg);
    else segments.push(seg);
  }

  return {
    shelterId: j?.shelter?.id != null ? String(j.shelter.id) : undefined,
    segments,
    connectorSegments,
    totalDistanceM: Math.trunc(Number(j?.total_distance_m ?? 0)),
    estimatedTimeMin: Number(j?.estimated_time_min ?? 0),
    totalRiskScore: Number(j?.total_risk_score ?? 0),
    riskPercent: Math.max(0, Math.min(100, Number(j?.risk_percent ?? 0))),
  };
}
